### ISSUE TITLE
Alpha affects linkForce differently than the other forces

### ISSUE THREAD

**Issue by asktree** [d3/d3-force#49]
Hi, I've been working on a project and noticed some weird behaviors around changes in alpha (namely due to alphaDecay) and the strength of links vs other forces; the lower the alpha, the weaker any linkForce is. I've set up a demo to show what I'm talking about.

In the demo, alpha decays as default and then resets to 1 once it reaches a threshold. This causes a "pulsating" behavior, where the graph expands as the links get weaker and then contracts once alpha is 1 and they're at full effect again. This happens regardless of whether I set my own strength function.

```
<!DOCTYPE html>
<meta charset="utf-8">
<body>
<script src="https://d3js.org/d3.v4.min.js"></script>
<script>

var svg = d3.select("body").append("svg")
	.attr("width", 400)
	.attr("height", 400)
	;
	
var simulation = d3.forceSimulation()
	.force("x", d3.forceX(200))
	.force("y", d3.forceY(200))
	.force("many", d3.forceManyBody())
	.force("link", d3.forceLink())
	;

simulation.force("many").strength(-400);

// It's more extreme with more iterations (and obviously higher strength)
// simulation.force("link").iterations(5);

simulation.force("link").distance(50);

mydata = [Object(1), Object(2), Object(3), 
		  Object(1), Object(2), Object(3),
		  Object(1), Object(2), Object(3),
		  Object(1), Object(2), Object(3),
		  Object(1), Object(2), Object(3),
		  Object(1), Object(2), Object(3)];
		  
var balls = svg.selectAll("circle")
	.data(mydata)

	.enter().append("circle")
	.attr("r", 5)
	;

//Here I just connect every ball to every other ball
links = []
balls.data().forEach( function (ball) {
	balls.data().forEach( function (ballbud) {
		if (ballbud != ball) {
			links.push({source: ball, target: ballbud});
		}
	});
});

simulation.nodes(balls.data());
simulation.force("link").links(links);

function update(){
	
	// Here's where I mess with the alpha -------------------------------------
	if (simulation.alpha() < .02) { 
		simulation.alpha(1);
	}
	
	balls
		.attr("cx", function (d) {return d.x; })
		.attr("cy", function (d) {return d.y; })
		;
}
simulation.on("tick", update);

</script>
```

I've tested all 3 of these forces (position, manybody, and link) against eachother in pairs and it's clearly the case that only link is affected; the "pulsating" behavior happens when only link and another of the forces are active, but not when the linkforce is disabled. It's also more evident that only link is affected in my actual project, where the graph is a lot more complex and populated. You can also just set alphaDecay to 0 and manually play with alpha, the result is the same.

Strangely the behavior seems different on one of Mike's examples: https://bl.ocks.org/mbostock/4600693
When you hold down the mouse on a node, everything kinda drifts apart (when I'd expect it to contract)

Hopefully this is helpful! I'm pretty new to d3 (huge fan so far) and JS in general (less of a fan so far); let me know if something in this snippet reflects some technical or stylistic misunderstandings. I'll poke around in the source and see if I can figure anything out.

**Comment by asktree** [comment]
I don't know if GitHub has a special request assignment feature, but I am working on fixing this.

**Comment by mbostock** [comment]
All of the forces are scaled by the simulation’s current alpha; that’s the intentional cooling of the forces. In link.js:

```js
l = (l - distances[i]) / l * alpha * strengths[i];
```

In x.js (and similarly y.js):

```js
node = nodes[i], node.vx += (xz[i] - node.x) * strengths[i] * alpha;
```

In manyBody.js:

```js
w = strengths[quad.data.index] * alpha / l;
```

If you instantaneously set the alpha to 1 when it goes below 0.02, I’d totally expect to see some pulsation. If you don’t want the simulation to stop permanently, maybe try setting alphaTarget to a non-zero value?

**Comment by asktree** [comment]
Sorry, I should be more clear. The issue I'm reporting on is not the slowdown or about any strange behaviors regarding instantly setting alpha to 1. Rather, what I'm referring to is that the forces are not equally scaled by the simulation's current alpha. In my first demo, the graph expands as alpha decays, which is what I was trying to show. I just set it to repeatedly jump back to alpha = 1 in order to demonstrate that motion. All the forces are cooling down, but linkForce seems to be cooling down more.

![image](https://cloud.githubusercontent.com/assets/12001874/17422149/e7084d98-5a74-11e6-9800-05779ba9c6c9.png)

For these examples I set alphaDecay to 0 and just set them each on a different alpha, and let them run for a while just to be sure they'd reached equilibrium. I also made the links more powerful to make it more pronounced, but the effect is there with any parameters (including defaults).

**Comment by mbostock** [comment]
I understand, but it’s also expected that as the simulation cools that the balance of forces changes. However, if you tinker with it and think you’ve found a way to improve the behavior, please let me know!

**Comment by asktree** [comment]
I ended up getting rid of this behavior by removing the velocity terms when calculating the positions of nodes. Basically, these terms end up getting more accurate as alpha is lower, so the force isn't as extreme because it doesn't over or under estimate (depending on the order you instantiate the forces) as much.

An alternative "fix" would be, instead of deleting the velocity terms, dividing them by alpha. I tried this and it also worked. However, I figured this way was pretty wonky and that it was better to just make links work like the other similar forces.

I realize this was probably a pretty low-on-the-priorities "fix", but my project is trying to display graphs with up to ~6000 links and it ends up being quite noticeable, all of the clusters in the graph shrink when you wake the simulation to drag stuff around and then start slowly ballooning back out again. I really like the alphaDecay feature so this bugged me a lot. 

I can see the merits of "preempting" nodes by checking where they're going to be, but in my opinion this behavior is more relevant when you're working with rigid forces like collision. Perhaps a forceRigidLink should be made, which simply prevents a node from ever moving more than a given distance away from a given point.

The other thing about preempting is that, without it, simulations really flip out when velocityDecay is 0. So with my change, this is true of forceLink; it's already true of manyBody and the positional force. It seemed pointless to prioritize support for velocityDecay(0) when the other forces already break with it. Using the alternative fix I described does not have this issue, though. Perhaps in the future they should all be changed in some way or another.

I initially thought that the problem was due to the fact that the old link.force did not factor in velocityDecay when preempting position. However, this doesn't seem to be the case because the behavior still existed when I just set velocityDecay to 0. Still seems like something to keep an eye out for in the future. Note that forceCollide also exhibits the alpha-inconsistency I tried to get rid of here, I just didn't try messing with it.

I went ahead and made a fork for your convenience and for the sake of my potentially being able to say I contributed to d3 :P
